### PR TITLE
Adds API for LRP introspection

### LINKED ISSUE #13156: Expose CLI debug interface for LRP
Discussed with @aditighag offline, when trying to debug LRP, currently there's no easily accessible interface to introspect the internal states of LRP manager and policies.

Proposal is to add a CLI interface (`cilium lrp list/get`) to dump internals of existing LRPs.

If this proposal makes sense, I can take a stab at the implementation.

### PR DESCRIPTION
**Pull request by Weil0ng**
Signed-off-by: Weilong Cui <cuiwl@google.com>

Fixes: #13156

### CODE REVIEW

**Comment by aditighag** [comment]
Thanks for the PR. I took a cursory glance at the changes. Can you add CLI output dump in the commit description?

**Comment by Weil0ng** [comment]
@aditighag I've added an example output in the commit msg, it is kinda arbitrary/hacky at this point, I'd like to get some input on what is the expected info/format to display :)

**Comment by aditighag** [comment]
Here is what I think would really help debug issues like "why isn't a local-redirect service entry created for an LRP?" - 

- LRP id (name and namespace) 

- Parsed LRP config which will include information like the LRP type (addressMatcher/svcMatcher), etc 
(https://github.com/cilium/cilium/blob/master/pkg/redirectpolicy/manager.go#L74)
(https://github.com/cilium/cilium/blob/master/pkg/redirectpolicy/redirectpolicy.go#L59)

- LRP frontends, pods (https://github.com/cilium/cilium/blob/master/pkg/redirectpolicy/manager.go#L66)

**Comment by Weil0ng** [comment]
> * LRP id (name and namespace)

This we already have with current implementation.

> * Parsed LRP config which will include information like the LRP type (addressMatcher/svcMatcher), etc
>   (https://github.com/cilium/cilium/blob/master/pkg/redirectpolicy/manager.go#L74)
>   (https://github.com/cilium/cilium/blob/master/pkg/redirectpolicy/redirectpolicy.go#L59)

Are you suggesting all the fields of LRPConfig should be listed?

> * LRP frontends, pods (https://github.com/cilium/cilium/blob/master/pkg/redirectpolicy/manager.go#L66)

Will add this.

**Comment by aditighag** [comment]
> 
> > * Parsed LRP config which will include information like the LRP type (addressMatcher/svcMatcher), etc
> >   (https://github.com/cilium/cilium/blob/master/pkg/redirectpolicy/manager.go#L74)
> >   (https://github.com/cilium/cilium/blob/master/pkg/redirectpolicy/redirectpolicy.go#L59)
> 
> Are you suggesting all the fields of LRPConfig should be listed?

We can start with the following ones - 
id, lrpType, serviceID (if present), backendSelector, frontendMappings, backendPorts

File: pkg/redirectpolicy/manager.go
Comment (nebril) [inline pkg/redirectpolicy/manager.go:797]: nitpick - can this be `RLock` instead of `Lock`?

**Review by nebril — APPROVED** [review]
LGTM! One nitpick left inline, but I don't think it's blocking.

File: api/v1/openapi.yaml
Comment (tklauser) [inline api/v1/openapi.yaml:563]: Nit: I think this should be `lrp` or similar instead, this API is doesn't seem related to services.
```suggestion
      - lrp
```

File: daemon/cmd/loadbalancer.go
Comment (tklauser) [inline daemon/cmd/loadbalancer.go]: Nit: stray empty line, looks like an unrelated change.

File: pkg/redirectpolicy/doc.go
Comment (tklauser) [inline pkg/redirectpolicy/doc.go]: Nit: drop this empty line, otherwise the comment above won't show up as package-level godoc.

File: pkg/redirectpolicy/redirectpolicy.go
Comment (tklauser) [inline pkg/redirectpolicy/redirectpolicy.go]: Please run `gofmt` on this change as per the GitHub action failure.

**Review by tklauser — APPROVED** [review]
Few minor nitpicks, mostly about formatting. Otherwise LGTM.

File: pkg/client/lrp.go
Comment (aanm) [inline pkg/client/lrp.go]: it should likely be `2016-2020`

**Review by aanm — COMMENTED** [review]
Nothing else besides what others point it out.

**Comment by brb** [comment]
@Weil0ng / @aditighag I'm wondering whether the same what the API provides can be achieved by using `kubectl get lrp`, `cilium service list` and some logging in the cilium-agent?

**Comment by aditighag** [comment]
> @Weil0ng / @aditighag I'm wondering whether the same what the API provides can be achieved by using `kubectl get lrp`, `cilium service list` and some logging in the cilium-agent?

There is no overlap with the information that I suggested we dump (https://github.com/cilium/cilium/pull/13327#issuecomment-700980248) and the commands you mentioned. `kubectl get lrp` will only show the CRD spec, which is further parsed by LRP handlers. I see a value in having both better logging and the ability to dump aggregated parsed/derived information in manager at _run time_. This would be similar to how we export some of the mapped internal state via `cilium service list`. Let me know what your thoughts are.

**Comment by brb** [comment]
Thanks for the explanation, it makes sense. My question arrived after seeing the example output in the commit msg:
```
Svc                   FrontendType       EndpointSelector
kube-system/lrp-svc   clusterIP + port   &LabelSelector{MatchLabels:map[string]string{k8s-app: node-local-dns,},MatchExpressions:[]LabelSelectorRequirement{},}
```

I was expecting that `kubectl get lrp -o yaml` would give me output with the exact details minus that it shows a desired state instead of a realized state.

**Review by brb — COMMENTED** [review]
Is `cilium lrp list` output going to be used not only by developers? If yes, then I think we should make it more pretty by formatting EndpointSelector.

Also, what does `Svc` stand for? Is it supposed to be an LRP name? If yes, then maybe `Name` makes more sense?

File: api/v1/openapi.yaml
Comment (joestringer) [reply api/v1/openapi.yaml:563]: I've been considering LRP to be a type of service, you have a virtual IP in front and the traffic to that IP is translated to a "real" IP behind. So `service` seems fine to me.

**Review by joestringer — CHANGES_REQUESTED** [review]
I didn't review; I just reviewed the reviews. I see comments about improving the CLI output format, and a scattering of other minor feedback including stuff like copyright. Let's not be afraid to make it clear when we are requesting that some feedback be addressed before merging.

**Comment by aditighag** [comment]
> Thanks for the explanation, it makes sense. My question arrived after seeing the example output in the commit msg:
> 
> ```
> Svc                   FrontendType       EndpointSelector
> kube-system/lrp-svc   clusterIP + port   &LabelSelector{MatchLabels:map[string]string{k8s-app: node-local-dns,},MatchExpressions:[]LabelSelectorRequirement{},}
> ```
> 
> I was expecting that `kubectl get lrp -o yaml` would give me output with the exact details minus that it shows a desired state instead of a realized state, and the realized state could be validated with `cilium service list`.

Right. The output details need to change. I called out some of the information that would be helpful to dump in the CLI output - 
https://github.com/cilium/cilium/pull/13327#issuecomment-700933648
https://github.com/cilium/cilium/pull/13327#issuecomment-700980248

@Weil0ng Based on our offline discussion, it seems like you are addressing these in this PR.

**Comment by aditighag** [comment]
> I didn't review; I just reviewed the reviews. I see comments about improving the CLI output format, and a scattering of other minor feedback including stuff like copyright. Let's not be afraid to make it clear when we are requesting that some feedback be addressed before merging.

Good point! I think I didn't choose the right option while posting the review comments. But I expect the CLI output format be fixed before this PR is merged.

**Comment by Weil0ng** [comment]
Thanks for the feedbacks, yes, I will make changes per @aditighag and others re formatting and others. I should have been more clear that the original output in the commit msg now is never meant to be merged as is, I just wanted to send out this PR soon to get feedbacks before polishing it. (For future reference) maybe there's a tag specific for this purpose I should use?

**Comment by joestringer** [comment]
@Weil0ng good question, I guess one common approach is to put "RFC" in the title. I think also while it's ideal to have the information in the commit messages (so we have a record in git), if the PR description doesn't point this out and doesn't contain any other information, then reviewers have to do a bit more digging to understand your intent, so I'd suggest trying to think "what do reviewers need to know about this PR to review?" and put that info in the PR description.

File: pkg/redirectpolicy/manager.go
Comment (aanm) [reply pkg/redirectpolicy/manager.go:797]: Let keep it as is, Aditi is changing the mutex from RWMutex to Mutex in https://github.com/cilium/cilium/pull/13362

File: daemon/cmd/lrp.go
Comment (christarazi) [inline daemon/cmd/lrp.go]: Nit:
```suggestion
	return NewGetLrpOK().WithPayload(getLRPs(h.rpm))
```

**Review by christarazi — CHANGES_REQUESTED** [review]
LGTM overall; one minor nit and `gofmt` the code as Tobias pointed out.

File: pkg/redirectpolicy/manager.go
Comment (Weil0ng) [reply pkg/redirectpolicy/manager.go:797]: Done.

File: daemon/cmd/loadbalancer.go
Comment (Weil0ng) [reply daemon/cmd/loadbalancer.go]: Done.

File: pkg/redirectpolicy/doc.go
Comment (Weil0ng) [reply pkg/redirectpolicy/doc.go]: Done.

File: pkg/redirectpolicy/redirectpolicy.go
Comment (Weil0ng) [reply pkg/redirectpolicy/redirectpolicy.go]: Done.

File: pkg/client/lrp.go
Comment (Weil0ng) [reply pkg/client/lrp.go]: Done.

File: daemon/cmd/lrp.go
Comment (Weil0ng) [reply daemon/cmd/lrp.go]: Done.

File: pkg/redirectpolicy/manager.go
Comment (Weil0ng) [reply pkg/redirectpolicy/manager.go:797]: oops, saw the second comment after refresh, reverted back to `Lock`.

File: pkg/redirectpolicy/redirectpolicy.go
Comment (christarazi) [inline pkg/redirectpolicy/redirectpolicy.go]: Nit:

```suggestion
	return  &models.LRPSpec{
		UID:              string(config.uid),
		Name:             config.id.Name,
		Namespace:        config.id.Namespace,
		FrontendType:     feType,
		LrpType:          lrpType,
		EndpointSelector: config.backendSelector.CachedLabelSelectorString,
	}
```

File: cilium/cmd/lrp.go
Comment (christarazi) [inline cilium/cmd/lrp.go]:
```suggestion
// Copyright 2020 Authors of Cilium
```

Comment (christarazi) [reply cilium/cmd/lrp.go]: There's a few occurrences of this for new files.

**Review by christarazi — APPROVED** [review]
LGTM. Minor nits to be addressed; does not require another review from me.

File: cilium/cmd/lrp.go
Comment (joestringer) [reply cilium/cmd/lrp.go]: Please address this before merging.

**Review by joestringer — CHANGES_REQUESTED** [review]
Minor comment, I didn't do a thorough review. Would be good to get one more review from @aditighag who knows this code the best, beyond that it should be good to merge.

**Comment by aditighag** [comment]
@Weil0ng Is this PR ready for review? I still don't see the expected CLI output in the commit description.

**Comment by Weil0ng** [comment]
> @Weil0ng Is this PR ready for review? I still don't see the expected CLI output in the commit description.

Sorry I haven't updated this yet, will mark as `wip` now and ask for review once I address all comments.

**Comment by aanm** [comment]
> > @Weil0ng Is this PR ready for review? I still don't see the expected CLI output in the commit description.
> 
> Sorry I haven't updated this yet, will mark as `wip` now and ask for review once I address all comments.

@Weil0ng I've marked the PR as draft. Feel free to mark it as ready for review once you have addressed all comments.

**Comment by Weil0ng** [comment]
> @Weil0ng Is this PR ready for review? I still don't see the expected CLI output in the commit description.

Updated, ready for another round :)

**Comment by aanm** [comment]
test-me-please

**Review by kkourt — APPROVED** [review]
for docs

File: cilium/cmd/lrp.go
Comment (christarazi) [reply cilium/cmd/lrp.go]: Don't think these have been addressed

**Comment by Weil0ng** [comment]
test-me-please

File: pkg/redirectpolicy/redirectpolicy.go
Comment (Weil0ng) [reply pkg/redirectpolicy/redirectpolicy.go]: Done.

File: cilium/cmd/lrp.go
Comment (Weil0ng) [reply cilium/cmd/lrp.go]: Done.

File: cilium/cmd/lrp_list.go
Comment (aditighag) [inline cilium/cmd/lrp_list.go:70]: What about address matcher LRPs?

**Review by aditighag — COMMENTED** [review]
Rest of the changes look good to me.

File: cilium/cmd/lrp_list.go
Comment (Weil0ng) [reply cilium/cmd/lrp_list.go:70]: the `Matching Service` column would be empty, see updated commit msg for an example (which has no backend).

**Comment by aditighag** [comment]
test-me-please

**Review by aditighag — APPROVED** [review]

**Comment by joestringer** [comment]
test-me-please

**Review by joestringer — APPROVED** [review]
Again, didn't review, just canceling my "changes requested" since others have now reviewed & changes appear to have been made to satisfy the feedback.

**Review by aanm — APPROVED** [review]

**Pull request merged**
